core/src/main/java/com/novoda/merlin/service/ConnectivityChangesForwarder.java

```diff
@@ -0,0 +1,68 @@
+package com.novoda.merlin.service;
+
+import com.novoda.merlin.NetworkStatus;
+import com.novoda.merlin.receiver.ConnectivityChangeEvent;
+import com.novoda.merlin.registerable.bind.BindListener;
+import com.novoda.merlin.registerable.connection.ConnectListener;
+import com.novoda.merlin.registerable.disconnection.DisconnectListener;
+
+class ConnectivityChangesForwarder {
+
+    private final NetworkStatusRetriever networkStatusRetriever;
+    private final DisconnectListener disconnectListener;
+    private final ConnectListener connectListener;
+    private final BindListener bindListener;
+    private final HostPinger hostPinger;
+
+    private NetworkStatus networkStatus;
+
+    ConnectivityChangesForwarder(NetworkStatusRetriever networkStatusRetriever,
+                                 DisconnectListener disconnectListener,
+                                 ConnectListener connectListener,
+                                 BindListener bindListener,
+                                 HostPinger hostPinger) {
+        this.networkStatusRetriever = networkStatusRetriever;
+        this.disconnectListener = disconnectListener;
+        this.connectListener = connectListener;
+        this.bindListener = bindListener;
+        this.hostPinger = hostPinger;
+    }
+
+    void notifyOfInitialNetworkStatus() {
+        if (networkStatus == null) {
+            bindListener.onMerlinBind(networkStatusRetriever.retrieveNetworkStatus());
```

Comment:
As far as I understand it the flow goes something like this.

- We notify of initial network state, setting the `networkStatus`. 
- `MerlinService` then registers for connectivity change events.
- An initial connectivity change event is sent upon registration.
- `MerlinService` forwards the connectivity change event to `ConnectivityChangesForwarder`, as this point the `NetworkStatus` is null, signalling that we do not have detailed information about the network state.
- At this point, we perform a ping. The ping determines that we have a valid network connection and notifies all listeners. We now store the `NetworkStatus` because we don't want to perform a ping more than is necessary.